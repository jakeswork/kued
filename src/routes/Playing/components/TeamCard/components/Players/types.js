import { arrayOf, shape, string } from 'prop-types';

export const propTypes = {
  classes: shape({}).isRequired,
  players: arrayOf(
    shape({
      charname: string,
      damageDone: string,
      deaths: string,
      healingDone: string,
      killingBlows: string,
      matchmaking_change: string,
      personal_change: string,
      realm: string,
      teamname: string,
      teamnamerich: string,
    }),
  ),
};

export const defaultProps = {
  players: [
    {
      charname: '',
      damageDone: '',
      deaths: '',
      healingDone: '',
      killingBlows: '',
      matchmaking_change: '',
      personal_change: '',
      realm: '',
      teamname: '',
      teamnamerich: '',
    },
  ],
};
