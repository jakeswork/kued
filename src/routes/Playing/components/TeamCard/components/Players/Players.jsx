import React, { Component } from 'react';
import { CircleSpinner } from 'react-spinners-kit';
import classNames from 'classnames';

import { propTypes, defaultProps } from './types';
import Text from '../../../../../../components/Text';
import Tooltip from '../../../../../../components/Tooltip';
import ErrorMessage from '../../../../../../components/ErrorMessage';

class Players extends Component {
  static propTypes = propTypes;

  static defaultProps = defaultProps;

  state = {
    players: [],
    loading: false,
    onlineIndicatorHovered: null,
    error: false,
  };

  componentDidMount() {
    this.setState({ loading: true });

    const { players } = this.props;
    const fetched = players.map(async (player) => {
      const data = await this.fetchPlayerData(player);

      return data;
    });

    return Promise.all(fetched).then(data => this.setState({ players: data, loading: false }));
  }

  fetchPlayerData = async (player) => {
    try {
      const response = await fetch(`http://localhost:4000/api/v1/warmane/player/${player.charname}/${player.realm}`);

      return await response.json();
    } catch (error) /* istanbul ignore next */ {
      // eslint-disable-next-line no-console
      console.error(new Error(error));

      return this.setState({ error: true });
    }
  }

  render() {
    const {
      players,
      loading,
      onlineIndicatorHovered,
      error,
    } = this.state;
    const { classes } = this.props;

    return (
      <div className={classes.root}>
        <CircleSpinner
          loading={loading}
          color="#fc3a3e"
          size={24}
        />
        <ErrorMessage error={error} />
        {
          (!loading && !error) && players.map((player, i) => (
            <div className={classes.playerWrapper} key={player.name}>
              <div
                className={classes.playerNameWrapper}
                onMouseEnter={() => this.setState({ onlineIndicatorHovered: i })}
                onMouseLeave={() => this.setState({ onlineIndicatorHovered: null })}
                data-test-id="playerWrapper"
              >
                <Text h5>
                  <span
                    className={classNames(
                      classes.onlineIndicator,
                      { [classes.online]: player.online },
                    )}
                  >
                    <Tooltip
                      visible={onlineIndicatorHovered === i}
                      style={{ left: -18, top: 18 }}
                      data-test-id="Tooltip"
                    >
                      { player.online ? 'Online' : 'Offline' }
                    </Tooltip>
                  </span>
                  { player.name }
                  &nbsp;
                  {
                    player.guild && (
                      <Text caption bold>
                        &lt;
                        { player.guild }
                        &gt;
                      </Text>
                    )
                  }
                </Text>
                <Text>
                  Level&nbsp;
                  { player.level }
                  &nbsp;
                  { player.race }
                  &nbsp;
                  { player.class }
                  &nbsp;
                  (
                  { player.spec }
                  )
                </Text>
              </div>
              <div className={classes.playerRatingWrapper}>
                {
                  player.pvpteams && (
                    player.pvpteams.constructor === Array
                      ? (
                        player.pvpteams.map(team => (
                          <div className={classes.playerRating} key={team.name}>
                            <Text caption>
                              { team.type }
                            </Text>
                            <Text bold>
                              { team.rating }
                            </Text>
                          </div>
                        ))
                      ) : (
                        <div className={classes.playerRating} key={player.pvpteams[1].name}>
                          <Text caption>
                            { player.pvpteams[1].type }
                          </Text>
                          <Text bold>
                            { player.pvpteams[1].rating }
                          </Text>
                        </div>
                      )
                  )
                }
              </div>
            </div>
          ))
        }
      </div>
    );
  }
}

export default Players;
