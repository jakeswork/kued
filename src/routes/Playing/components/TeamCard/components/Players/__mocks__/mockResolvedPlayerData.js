const player = {
  spec: 'Feral Combat',
  name: 'Apparen',
  realm: 'Icecrown',
  online: false,
  level: '80',
  guild: 'Test',
  race: 'Night Elf',
  class: 'Druid',
  pvpteams: [{
    type: '2v2', name: 'Hello', rating: '2181', rank: '18',
  }, {
    type: '3v3', name: 'apparent', rating: '336', rank: '100+',
  }],
};

export default [
  player,
  {
    ...player,
    name: 'Other',
    pvpteams: {
      1: {},
    },
  },
];
