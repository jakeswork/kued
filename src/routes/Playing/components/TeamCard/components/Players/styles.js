export default theme => ({
  root: {
    display: 'flex',
    justifyContent: 'center',
    padding: '16px 0',
    flexDirection: 'column',
    alignItems: 'center',
    flexWrap: 'wrap',
  },
  playerWrapper: {
    display: 'flex',
    justifyContent: 'space-between',
    width: '100%',
    alignItems: 'center',
    margin: '8px 0',
  },
  playerNameWrapper: {
    textAlign: 'left',
    position: 'relative',
  },
  onlineIndicator: {
    display: 'inline-block',
    position: 'relative',
    marginRight: 8,
    width: 12,
    height: 12,
    borderRadius: '100%',
    backgroundColor: theme.colorRed,
  },
  playerRating: {
    margin: '0 8px',
  },
  playerRatingWrapper: {
    display: 'flex',
    justifyContent: 'space-between',
  },
  online: {
    backgroundColor: theme.colorGreen,
  },
});
