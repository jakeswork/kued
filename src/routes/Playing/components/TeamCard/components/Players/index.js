import withStyles from 'react-jss';

import Players from './Players';
import styles from './styles';

export default withStyles(styles)(Players);
