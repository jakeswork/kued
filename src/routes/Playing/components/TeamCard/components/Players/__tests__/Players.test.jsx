import React from 'react';
import { shallow } from 'enzyme';

import Players from '../Players';
import mockResolvedPlayerData from '../__mocks__/mockResolvedPlayerData';

const defaultProps = {
  classes: {},
  players: [
    {
      charname: 'Test',
      damageDone: '123',
      deaths: '1',
      healingDone: '123',
      killingBlows: '1',
      matchmaking_change: '123',
      personal_change: '123',
      realm: 'Blackrock',
      teamname: 'Test',
      teamnamerich: 'Test',
    },
  ],
};

describe('The Players instance', () => {
  describe('When rendered with default props', () => {
    it('should render without throwing an error', () => {
      const wrapper = shallow(<Players {...defaultProps} />);

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('When the players are loaded', () => {
    it('should render player information', () => {
      const wrapper = shallow(<Players {...defaultProps} />);

      wrapper.setState({ loading: false, players: mockResolvedPlayerData });

      const playerWrapper = wrapper.find({ 'data-test-id': 'playerWrapper' });

      expect(playerWrapper.exists()).toBe(true);
    });

    describe('When a player card is hovered', () => {
      it('Should display a tooltip describing if they are online or not', () => {
        const wrapper = shallow(<Players {...defaultProps} />);

        wrapper.setState({ loading: false, players: mockResolvedPlayerData });

        const playerWrapper = wrapper.find({ 'data-test-id': 'playerWrapper' }).first();

        playerWrapper.simulate('mouseenter');

        const Tooltip = wrapper.find({ 'data-test-id': 'Tooltip' }).first();

        expect(Tooltip.props().children).toEqual('Offline');

        playerWrapper.simulate('mouseleave');

        const onlineMock = [{ ...mockResolvedPlayerData[0], online: true }];

        wrapper.setState({ players: onlineMock });

        playerWrapper.simulate('mouseenter');

        const newTooltip = wrapper.find({ 'data-test-id': 'Tooltip' });

        expect(newTooltip.props().children).toEqual('Online');
      });
    });
  });

  describe('When fetching data', () => {
    it('Should update the loading state and set the players', () => {
      const newProps = {
        ...defaultProps,
        players: [{
          ...defaultProps.players[0],
          charname: 'Test',
          realm: 'Icecrown',
        }],
      };
      const mockSuccessResponse = {};
      const mockJsonPromise = Promise.resolve(mockSuccessResponse);
      const mockFetchPromise = Promise.resolve({
        json: () => mockJsonPromise,
      });
      jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise);

      const wrapper = shallow(<Players {...newProps} />);

      wrapper.setState({ players: mockResolvedPlayerData });

      expect(global.fetch).toHaveBeenCalledTimes(1);
      expect(global.fetch).toHaveBeenCalledWith(`http://localhost:4000/api/v1/warmane/player/${newProps.players[0].charname}/${newProps.players[0].realm}`);
    });
  });
});
