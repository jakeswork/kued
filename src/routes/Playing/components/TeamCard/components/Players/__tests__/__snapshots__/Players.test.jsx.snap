// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Players instance When rendered with default props should render without throwing an error 1`] = `
<div>
  <Component
    color="#fc3a3e"
    loading={true}
    size={24}
    sizeUnit="px"
  />
  <Jss(ErrorMessage)
    error={false}
  />
</div>
`;
